File: src/02.string/07.match/07.214.ts
export function shortestPalindrome(s: string): string {
    const n = s.length;
    let endIndex = n;
    for (let i = n - 1; i >= 0; i--) {
        let j = i;
        let k = 0;
        let isCompleted = true;
        for (; k <= j; k++, j--) {
            if (s[k] !== s[j]) {
                isCompleted = false;
                break;
            }
        }

        if (isCompleted) {
            endIndex = i + 1;
            break;
        }
    }

    if (endIndex >= n) {
        return s;
    }

    let res = "";
    for (let i = n - 1; i >= endIndex; i--) {
        res += s[i];
    }

    return res + s;
}

File: src/02.string/07.match/07.214.go
package match

import "strings"

func ShortestPalindrome(s string) string {
	n := len(s)
	end := n

	for k := n - 1; k >= 0; k-- {
		i, j, isFailed := 0, k, true
		for ; i <= j; i, j = i+1, j-1 {
			if s[i] != s[j] {
				isFailed = false
				break
			}
		}
		if isFailed {
			end = k + 1
			break
		}
	}

	if end >= n {
		return s
	}

	res := &strings.Builder{}
	for i := n - 1; i >= end; i-- {
		res.WriteByte(s[i])
	}

	return res.String() + s
}
